mono/sgen/sgen-marksweep.c

```diff
@@ -1811,18 +1836,22 @@ major_start_major_collection (void)
 		free_block_lists [MS_BLOCK_FLAG_REFS][i] = NULL;
 	}
 
-	if (lazy_sweep)
-		binary_protocol_sweep_begin (GENERATION_OLD, TRUE);
+	if (lazy_sweep && concurrent_sweep) {
+		SgenThreadPoolJob *job = sweep_blocks_job;
+		if (job)
```

Comment:
sweep_blocks_job is created before sweep_finish, which we still wait for above (major_finish_sweep_checking). After the end of sweep, if we don't have sweep_blocks_job set, it means that it has already been run.